k2/csrc/connect.h

```diff
@@ -0,0 +1,111 @@
+
+#ifndef K2_CSRC_CONNECT_H_
+#define K2_CSRC_CONNECT_H_
+
+#include <vector>
+
+#include "k2/csrc/fsa.h"
+
+namespace k2 {
+/*
+  Removes states that are not accessible (from the start state) or are not
+  co-accessible (i.e. that cannot reach the final state), and ensures that
+  if the FSA admits a topological sorting (i.e. it contains no cycles except
+  self-loops), the version that is output is topologically sorted.  This
+  is not a stable sort, i.e. states may be renumbered even for top-sorted
+  input.
+
+  Notes:
+    - If the input FSA (`fsa_in`) admitted a topological sorting, the output
+      FSA (`fsa_out`) will be topologically sorted. If `fsa_in` is not
+      topologically sorted but is acyclic, `fsa_out` will also be topologically
+      sorted. TODO(Dan): maybe just leave in the same order as `fsa_in`??
+      (Current implementation may **renumber** the state)
+    - If `fsa_in` was deterministic, `fsa_out` will be deterministic; same for
+      epsilon free, obviously.
+    - `fsa_out` will be arc-sorted (arcs sorted by label). TODO(fangjun): this
+       has not be implemented.
+ */
+class Connection {
+ public:
+  /* Lightweight constructor that just keeps const references to the input
+     parameters.
+     @param [in] fsa_in  The input FSA
+  */
+  explicit Connection(const Fsa &fsa_in) : fsa_in_(fsa_in) {}
+
+  /*
+    Do enough work that know now much memory will be needed, and output
+    that information
+        @param [out] fsa_size   The num-states and num-arcs of the output FSA
+                                will be written to here
+  */
+  void GetSizes(Array2Size<int32_t> *fsa_size);
+
+  /*
+    Finish the operation and output the connected FSA to `fsa_out` and
+    arc mapping information to `arc_map` (if provided).
+    @param [out] fsa_out   The output FSA; Will be connected;
+                           Must be initialized; search for 'initialized
+                           definition' in class Array2 in array.h for meaning.
+    @param [out] arc_map   If non-NULL, will output a map from the arc-index
+                           in `fsa_out` to the corresponding arc-index in
+                           `fsa_in`.
+                           If non-NULL, at entry it must be allocated with
+                           size num-arcs of `fsa_out`, e.g. `fsa_out->size2`.
+
+    @return   The return status indicates whether topological sorting
+        was successful; if true, the result is top-sorted.  The only situation
+        it might return false is when the input had cycles that were not self
+        loops; such FSAs do not admit a topological sorting.
+
+        Caution: true return status does not imply that the returned FSA is
+        nonempty.
+   */
+  bool GetOutput(Fsa *fsa_out, int32_t *arc_map = nullptr);
+
+ private:
+  const Fsa &fsa_in_;
+
+  bool is_acyclic_;  // if the input FSA is acyclic
+  // indicates that there's no any accessible & co-accessible state
```

Comment:
this comment is unclear.  I'd write: // if true, there is no state in the input FSA that is both accessible and co-accessible.
(Assuming that's what you mean!  Or do you mean there exists at least one state that is either not accessible or not co-accessible?)

Reply:
Your first version, updated, thanks!